File: tests/Generator/resource/csharp/complex/CollectionDefinitionType.cs
using System.Text.Json.Serialization;

namespace TypeAPI.Model;

/// <summary>
/// Base collection type
/// </summary>
[JsonPolymorphic(TypeDiscriminatorPropertyName = "type")]
[JsonDerivedType(typeof(ArrayDefinitionType), typeDiscriminator: "array")]
[JsonDerivedType(typeof(MapDefinitionType), typeDiscriminator: "map")]
public abstract class CollectionDefinitionType : DefinitionType
{
    [JsonPropertyName("schema")]
    public PropertyType? Schema { get; set; }

    [JsonPropertyName("type")]
    public new string? Type { get; set; }

}

File: tests/Generator/resource/rust/complex/collection_definition_type.rs
use serde::{Serialize, Deserialize};
use array_definition_type::ArrayDefinitionType;
use map_definition_type::MapDefinitionType;
use definition_type::DefinitionType;
use property_type::PropertyType;

// Base collection type
#[derive(Serialize, Deserialize)]
pub struct CollectionDefinitionType {
    #[serde(rename = "deprecated")]
    deprecated: Option<bool>,

    #[serde(rename = "description")]
    description: Option<String>,

    #[serde(rename = "type")]
    _type: Option<String>,

    #[serde(rename = "schema")]
    schema: Option<PropertyType>,

}

